pkg/clustermanager/controller.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustermanager
+
+import (
+	"context"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	lcv1alpha1 "github.com/kubestellar/kubestellar/pkg/apis/logicalcluster/v1alpha1"
+	edgeclient "github.com/kubestellar/kubestellar/pkg/client/clientset/versioned"
+)
+
+const ()
+
+type triggeringKind string
+
+const (
+	triggeringKindLogicalCluster      triggeringKind = "LogicalCluster"
+	triggeringKindClusterProviderDesc triggeringKind = "ClusterProviderDesc"
+	controllerName                                   = "logicalcluster-manager"
+)
+
+type queueItem struct {
+	triggeringKind triggeringKind
+	key            string
+}
+
+type controller struct {
+	context                 context.Context
+	clientset               edgeclient.Interface
+	k8sClientset            *kubeclient.Clientset
+	logger                  logr.Logger
+	queue                   workqueue.RateLimitingInterface
+	logicalClusterInformer  cache.SharedIndexInformer
+	clusterProviderInformer cache.SharedIndexInformer
+	providers               map[string]*provider
+	lock                    sync.Mutex
+}
+
+// NewController returns logicalcluster-manager controller
+func NewController(
+	context context.Context,
```

Comment:
Not a blocker: naming a parameter `context` shadows the imported package `context`.  I commonly see the parameter named `ctx` instead.

Reply:
Handled by PR #816